### PR TITLE
Scripts

### PR DESCRIPTION
**Pull request by NiklasGustafsson**
Addressing #770 and #765.

### CODE REVIEW

**Comment by NiklasGustafsson** [comment]
@kaiidams -- this is in response to your recent issues. If you have some time, please take a look at it.

File: src/TorchSharp/JIT/CompilationUnit.cs [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                    if (!objs.All(o => typeof(Tensor).IsAssignableFrom(o.GetType()))) {
+                        throw new NotImplementedException($"CompilationUnit.{name}() is not yet taking non-tensors as input arguments");
+                    }
+
+                    IntPtr[] ptrArray = null;
+                    sbyte typeCode = 0;
+
+                    using (var parray = new PinnedArray<IntPtr>()) {
+
+                        var tensors = objs.Select(o => (Tensor)o).ToArray();
```
Comment (tarekgh) [inline src/TorchSharp/JIT/CompilationUnit.cs:86]: do we need to do that? 

is it possible we cast the array items to `Tensot` inside the loop 

```
for (var i = 0; i < tensors.Length; i++) tensorRefs[i] = tensors[i].Handle;
```

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 2]
```diff
@@ -332,6 +337,57 @@ public object forward(params object[] objs)
...
+                    if (!objs.All(o => typeof(Tensor).IsAssignableFrom(o.GetType()))) {
+                        throw new NotImplementedException($"ScriptModule.{name}() is not yet taking non-tensors as input arguments");
+                    }
+
+                    IntPtr[] ptrArray = null;
+                    sbyte typeCode = 0;
+
+                    using (var parray = new PinnedArray<IntPtr>()) {
+
+                        var tensors = objs.Select(o => (Tensor)o).ToArray();
```
Comment (tarekgh) [inline src/TorchSharp/JIT/ScriptModule.cs:362]: >var tensors = objs.Select(o => (Tensor)o).ToArray(); [](http://example.com/codeflow?start=24&length=52)

same previous comment.

**Review by tarekgh — DISMISSED** [review]

**Review by kaiidams — APPROVED** [review]

**Comment by NiklasGustafsson** [comment]
@tarekgh, @kaiidams -- I need to bother you with a re-review. I've added limited support for additional types in TorchScript, and that code needs some checking. Specifically, it's the passing of TensorOrScalar to and from native code, as well as the processing of arguments and return values on either side.

File: src/Native/LibTorchSharp/THSJIT.cpp [hunk 3]
```diff
@@ -145,45 +155,174 @@ JITMethod THSJIT_Module_get_method(const JITModule module, const char* name)
...
+
+        auto list = result.toList();
+        TensorOrScalar* output = allocator(list.size());
+
+        for (int i = 0; i < list.size(); ++i)
+        {
+            output[i].Handle = -1;
+            c10::IValue value = list[i];
+
+            if (value.isTensor())
```
Comment (tarekgh) [inline src/Native/LibTorchSharp/THSJIT.cpp:208]: >value [](http://example.com/codeflow?start=16&length=5)

is it possible the list can have tuple items? I mean is it possible we can run into the case that have value.isTuple() is true?

File: src/Native/LibTorchSharp/THSJIT.cpp [hunk 4]
```diff
@@ -145,45 +155,174 @@ JITMethod THSJIT_Module_get_method(const JITModule module, const char* name)
...
+        auto result = (*module)->forward(toIValue(tensorPtrs, length));
+        ReturnHelper(result, allocator, typeCode);
+    )
+}
+
+void THSJIT_Module_invoke(const JITModule module, const char* name, const TensorOrScalar* tensorPtrs, const int length, TensorOrScalar* (*allocator)(size_t length), int8_t* typeCode)
+{
+    *typeCode = 0;
+
+    //CATCH(
```
Comment (tarekgh) [inline src/Native/LibTorchSharp/THSJIT.cpp:310]: >CATCH [](http://example.com/codeflow?start=6&length=5)

why the catch is commented here?

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 5]
```diff
@@ -271,53 +275,169 @@ private Type GetType(Type type)
...
+                /// 3. Array (Python list) of Tensor / Scalar
+                ///
+                /// For returned types, if the number of values returned in a tuple is greaterh than 5, it is returned as an array, instead.
+                /// If a tuple contains both tensors and scalars, it is returned as an object[].
+                /// </remarks>
+                public object invoke(string name, params object[] objs)
+                {
+                    if (String.IsNullOrEmpty(name)) throw new ArgumentNullException("method name");
+
+                    //if (!objs.All(o => typeof(Tensor).IsAssignableFrom(o.GetType()))) {
```
Comment (tarekgh) [inline src/TorchSharp/JIT/ScriptModule.cs:352]: >// [](http://example.com/codeflow?start=20&length=2)

do you still need to keep this comment?

File: src/TorchSharp/JIT/ScriptModule.cs [hunk 6]
```diff
@@ -271,53 +275,169 @@ private Type GetType(Type type)
...
+                            tensorRefs[idx].TypeCode = 3;
+                            break;
+                        case long l:
+                            tensorRefs[idx].Handle = ((Scalar)l).Handle;
+                            tensorRefs[idx].TypeCode = 1;
+                            // The MacOS version of Clang doesn't like the use of int64_t, so pass as a Scalar instance, instead.
+                            //tensorRefs[idx].Handle = (IntPtr)l;
+                            //tensorRefs[idx].TypeCode = 4;
+                            break;
+                        default:
```
Comment (tarekgh) [inline src/TorchSharp/JIT/ScriptModule.cs:408]: nit: align this block?

File: src/TorchSharp/JIT/ScriptModule.cs (hunk 5 above)
Comment (NiklasGustafsson) [reply src/TorchSharp/JIT/ScriptModule.cs:352]: Nope. Gone!

File: src/Native/LibTorchSharp/THSJIT.cpp (hunk 3 above)
Comment (NiklasGustafsson) [reply src/Native/LibTorchSharp/THSJIT.cpp:208]: Yes, and we'll reject those for now.

**Review by tarekgh — DISMISSED** [review]

**Pull request merged**